### PR TITLE
Free Cantrips

### PR DESCRIPTION
**Pull request by inkcorvid**
An event handler has been created (subscribed on mod load) for when spell slots are consumed, which will refund level 0 cantrips cast by wizards and sorcerers. It handles multiclassing and metamagic gracefully: only cantrip slots will ever be refunded, and only the cantrips wizards prepared at the start of the adventuring day are unlimited.

An ImpactScript has been created for the spell Disrupt Undead. It works as the spell description describes.

Cantrips no longer contribute to weave corruption or repair. Spell level is passed through the functions in the x2 spellhook into loi_weave. Weave state changes only happen if that is > 0

The unuse script tb_spellhook has been deleted.

A SpellFizzle function has been created in loi_weave for when dead magic causes a spell to fail, which looks up the casting animation type and plays the appropriate "fizzle" VFX, as if the spell had been interrupted. It is called in the x2 spellhook alongside the dead magic function.

### CODE REVIEW

File: src/scripts/te_sp_disruptud.nss [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        if(!GetIsReactionTypeFriendly(oTarget) && GetRacialType(oTarget) == RACIAL_TYPE_UNDEAD)
+        {
+            //Fire cast spell at event for the specified target
+            SignalEvent(oTarget, EventSpellCastAt(OBJECT_SELF, 1016));
+            
+            //Make SR Check
+            if(!MyResistSpell(OBJECT_SELF, oTarget))
+            {
+                //Set damage effect
```
Comment (siobhan-buck) [inline src/scripts/te_sp_disruptud.nss:39]: effect eVis = EffectVisualEffect(VFX_IMP_DEATH); 
is only needed in this if block, so it should be moved here to save on executions

File: src/scripts/te_modload.nss [hunk 2]
```diff
@@ -223,4 +223,6 @@ void main()
     NWNX_Events_SubscribeEvent("NWNX_ON_DM_GIVE_LEVEL_AFTER","event_dm");
     NWNX_Events_SubscribeEvent("NWNX_ON_DM_GIVE_ALIGNMENT_AFTER","event_dm");
     NWNX_Events_SubscribeEvent("NWNX_ON_CLIENT_DISCONNECT_BEFORE","event_leave");
+
+    NWNX_Events_SubscribeEvent("NWNX_ON_DECREMENT_SPELL_COUNT_AFTER", "loi_refund_spell");
```
Comment (siobhan-buck) [inline src/scripts/te_modload.nss:227]: so pretty~

File: src/scripts/tb_spellhook.nss
Comment (siobhan-buck) [inline src/scripts/tb_spellhook.nss:1]: die die DIE

**Review by siobhan-buck — CHANGES_REQUESTED** [review]
in general, looks great! I have a few style / optimizations comments

File: src/scripts/te_sp_disruptud.nss (hunk 1 above)
Comment (inkcorvid) [reply src/scripts/te_sp_disruptud.nss]: Moved! And corrected a little typo that would have prevented the bigger animation from playing on crit.

File: src/scripts/x2_inc_spellhook.nss [hunk 3]
```diff
@@ -344,13 +344,19 @@ int X2PreSpellCastCode()
...
+        // so we can exclude cantrips from corrupting the Weave
+        int nSpellLevel = GetLastSpellLevel();
+
         // Only apply wild and dead magic to PCs.
         // We don't want environmental effects or DM events to break.
-        if (X2DeadmagicZone(OBJECT_SELF) == FALSE)
+        if (X2DeadmagicZone(OBJECT_SELF, nSpellLevel) == FALSE)
         {
+            int nSpellID = GetSpellId();
+            SpellFizzle(OBJECT_SELF, nSpellID);
```
[the remark this answers is not in the document]
Comment (inkcorvid) [reply src/scripts/x2_inc_spellhook.nss]: Refactored! X2DeadmagicZone now takes a spell ID, and the fizzle function is called from within it just before it returns FALSE.

File: src/scripts/loi_refund_spell.nss [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            // get the index of the last slot
+            int nSlot = GetMemorizedSpellCountByLevel(oPC, CLASS_TYPE_WIZARD, 0);
+
+            // loop down the cantrips section of the spellbook
+            // if you find one with the same id as cast, refund it
+            // index bounds 0 <= nIndex < GetMemorizedSpellCountByLevel()
+            while (nSlot > 0)
+            {
+                nSlot -= 1;
```
[the remark this answers is not in the document]
Comment (inkcorvid) [reply src/scripts/loi_refund_spell.nss]: Changed to use a for loop

File: src/scripts/loi_refund_spell.nss [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    int nSpell = StringToInt(NWNX_Events_GetEventData("SPELL_ID"));
+    int nSL = SFGetInnateSpellLevelByClass(nSpell, CLASS_TYPE_WIZARD);
+    int nMetamagic = StringToInt(NWNX_Events_GetEventData("METAMAGIC"));
+
+    int nCasterClassIndex = StringToInt(NWNX_Events_GetEventData("CLASS"));
+    int nCasterClass = GetClassByPosition(nCasterClassIndex + 1);
+
+    if (nSL == 0 && nMetamagic == METAMAGIC_NONE)
+    {
+
```
[the remark this answers is not in the document]
Comment (inkcorvid) [reply src/scripts/loi_refund_spell.nss]: Moved!

File: src/scripts/loi_refund_spell.nss
[the remark this answers is not in the document]
Comment (inkcorvid) [reply src/scripts/loi_refund_spell.nss:1]: Renamed!

**Comment by inkcorvid** [comment]
This no longer works for some reason. Closing until I can get it to work

**Comment by inkcorvid** [comment]
Have tested it, and it works now!

**Review by siobhan-buck — APPROVED** [review]

**Pull request merged**